src/server/kernel/wsurface.cpp

```diff
@@ -158,6 +166,26 @@ void WSurfacePrivate::updateBuffer()
     setBuffer(buffer);
 }
 
+void WSurfacePrivate::updatePreferredBufferScale()
+{
+    if (explicitPreferredBufferScale > 0)
+        return;
+
+    float maxScale = 1.0;
+    for (auto o : outputs)
+        maxScale = std::max(o->scale(), maxScale);
+    preferredBufferScale = qCeil(maxScale);
```

Comment:
Why not qRound ？

Reply:
Must ensure wl_surface's buffer scale greater than wl_output's scale.